### ISSUE TITLE
Pretty Print JSON with Mantle

### ISSUE THREAD

**Issue by istx25** [Mantle/Mantle#741]
It's possible that I misunderstood the benefits of using Mantle, but this is what I'm trying to do:

- Serialise `MTLObject` objects into JSON (w/ snake-case key names).
- Convert the serialised JSON dictionary into a "Pretty Printed" JSON string.
- Display the Pretty Printed JSON String in a `UITextView` and/or `UILabel`.

Why?

I'm working on a research study tool and our API isn't ready yet. At the end of each trial, the facilitator emails the data collected (or shares it to Notes.app) for processing at a later time.

Here is the object which runs the show:

```swift
import Mantle

final class Trial: MTLModel, MTLJSONSerializing {
    
    // MARK: Properties
    var duration: TrialDuration?
    var scene: TrialScene?
    var observations: [Observation]?
    
    // MARK: Mantle
    static func JSONKeyPathsByPropertyKey() -> [NSObject : AnyObject]! {
        return ["duration": "duration",
                "scene": "scene",
                "observations": "observations"]
    }
}
```

- `TrialDuration` takes in two `NSDate` objects which represent when the trial started and ended.
- `TrialScene` collects information about the participant and location of the trial.
     - A `String?` property for the participant's unique identifier.
     - A `String?` to describe the location of the trial in a paragraph or two.
     - A `NSData?` object to represent a `UIImage` of the trial location.
- A collection of `Observation` objects which are created when a change is observed.
     - `Observation` is specific to the research study, can't share it.
     - Similar structure to `Trial` (multiple objects [with properties] contained in one main object).
     - Has multiple `NSDate` objects.
     - Has multiple `Set<String>` objects.
     - Has a `(String, Array<Int>)` tuple.
     - Has a bunch of `String?` properties.

I tried to use `NSJSONSerialization` but I receive a crasher.

```swift
let data = try! NSJSONSerialization.dataWithJSONObject(trial, options: .PrettyPrinted)
let string = NSString(data: data, encoding: NSUTF8StringEncoding)

/// '*** +[NSJSONSerialization dataWithJSONObject:options:error:]: 
/// Invalid top-level type in JSON write'
```

What's the best way to do this with Mantle? 
If it's not possible, can you point me in the right direction?

Cheers!

💃 😸

**Comment by conradev** [comment]
You need to transform the `trial` into a JSON object before passing it into `NSJSONSerialization`.

Call `+[MTLJSONAdapter JSONDictionaryFromModel:error:]` with the trial and try serializing the result of that method

Let me know if that works! ✨

**Comment by istx25** [comment]
Thanks for responding, @conradev! 💖

```swift
let dictionary = try? MTLJSONAdapter.JSONDictionaryFromModel(self.trial, error: ())
let data = try! NSJSONSerialization.dataWithJSONObject(dictionary!, options: .PrettyPrinted)
let string = NSString(data: data, encoding: NSUTF8StringEncoding)

/// 'Invalid type in JSON write (<Project_Name>.Observation)'
```

Progress! But I still get an uncaught exception at the `let data`line. I looked up [which types are actually supported](http://www.tutorialspoint.com/json/json_data_types.htm) by JSON and I'm guessing the tuple is the culprit.

```swift
import Mantle

final class ObservableLevel: MTLModel, MTLJSONSerializing {
    
    // MARK: Properties
    var someLevel: (String, [Int])?
    
    // MARK: Mantle
    static func JSONKeyPathsByPropertyKey() -> [NSObject : AnyObject]! {
        return ["someLevel": "some_level"]
    }
}
```
- `ObservableLevel` is on a property inside the `Observation` object.

I'm guessing this is where I would need to use a `NSValueTransformer`. Correct? I'm still quite confused by them. How would I use a transformer in this situation?

**Comment by conradev** [comment]
Correct. You can't represent a tuple in JSON.

The implementation of the transformer depends on how you want to go about representing `someLevel` in JSON. I have trouble understanding how it would be structured best it because I don't quite get the model. Did you mean to have something like this?

```swift
import Mantle

final class ObservableLevel: MTLModel, MTLJSONSerializing {
    // MARK: Properties
    var name: String?
    var levels: [Int]?
}
```

It's rare that I see a model with a singular tuple as a property

**Comment by istx25** [comment]
@conradev: That makes sense. I changed `ObservableLevel` to have two properties instead of a tuple. I tried it again and (still) got the same exception.

```
2016-07-24 15:47:03.896 <Project_Name>[34824:2904145] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: 'Invalid type in JSON write (<Project_Name>.Observation)'
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000104fabd85 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x0000000106d4fdeb objc_exception_throw + 48
	2   CoreFoundation                      0x0000000104fabcbd +[NSException raise:format:] + 205
	3   Foundation                          0x00000001054a0d37 _writeJSONValue + 698
	4   Foundation                          0x00000001054a4b55 ___writeJSONArray_block_invoke + 132
	5   CoreFoundation                      0x0000000104efc374 __NSArrayEnumerate + 596
	6   Foundation                          0x00000001054a438d _writeJSONArray + 330
	7   Foundation                          0x00000001054a0ca5 _writeJSONValue + 552
	8   Foundation                          0x00000001054a4992 ___writeJSONObject_block_invoke + 226
	9   CoreFoundation                      0x0000000104f11ac5 __65-[__NSDictionaryI enumerateKeysAndObjectsWithOptions:usingBlock:]_block_invoke + 85
	10  CoreFoundation                      0x0000000104f119cd -[__NSDictionaryI enumerateKeysAndObjectsWithOptions:usingBlock:] + 237
	11  Foundation                          0x00000001054a411c _writeJSONObject + 436
	12  Foundation                          0x00000001054a0c09 _writeJSONValue + 396
	13  Foundation                          0x00000001054a0a29 -[_NSJSONWriter dataWithRootObject:options:error:] + 124
	14  Foundation                          0x00000001054a344a +[NSJSONSerialization dataWithJSONObject:options:error:] + 333
	15  <Project_Name>                      0x00000001047b71f1 _TFC6<Project_Name>20ExportViewController18shareButtonPressedfCS_16ReversibleButtonT_ + 1441
	16  <Project_Name>                      0x00000001047b759a _TToFC6<Project_Name>20ExportViewController18shareButtonPressedfCS_16ReversibleButtonT_ + 58
	17  UIKit                               0x00000001057d5a8d -[UIApplication sendAction:to:from:forEvent:] + 92
	18  UIKit                               0x0000000105948e67 -[UIControl sendAction:to:forEvent:] + 67
	19  UIKit                               0x0000000105949143 -[UIControl _sendActionsForEvents:withEvent:] + 327
	20  UIKit                               0x0000000105948263 -[UIControl touchesEnded:withEvent:] + 601
	21  UIKit                               0x000000010584899f -[UIWindow _sendTouchesForEvent:] + 835
	22  UIKit                               0x00000001058496d4 -[UIWindow sendEvent:] + 865
	23  UIKit                               0x00000001057f4dc6 -[UIApplication sendEvent:] + 263
	24  UIKit                               0x00000001057ce553 _UIApplicationHandleEventQueue + 6660
	25  CoreFoundation                      0x0000000104ed1301 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
	26  CoreFoundation                      0x0000000104ec722c __CFRunLoopDoSources0 + 556
	27  CoreFoundation                      0x0000000104ec66e3 __CFRunLoopRun + 867
	28  CoreFoundation                      0x0000000104ec60f8 CFRunLoopRunSpecific + 488
	29  GraphicsServices                    0x000000010aa51ad2 GSEventRunModal + 161
	30  UIKit                               0x00000001057d3f09 UIApplicationMain + 171
	31  <Project_Name>                      0x00000001047b3602 main + 114
	32  libdyld.dylib                       0x000000010859b92d start + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
(lldb) 
```

Are these types/objects automatically transformed, or do I need to do this myself?

- `NSDate`
- `Set<String>`

Those are the only two which stand out to me. The rest are `String` and `[Int]`. It's at times like this where it feels like my debugging skills are futile, haha. Do you have any tricks up your sleeve that will help me find what's causing the problem?

Looking forward to reading your response.

**Comment by istx25** [comment]
Okay... some progress...

- I wrote a custom `NSValueTransformer` for `NSDate`.

```swift
static func dateJSONTransformer() -> NSValueTransformer {
    let forwardBlock: MTLValueTransformerBlock? = { (value, success, error) in
        return dateFormatter.dateFromString(value as! String!)
    }
        
    let reverseBlock: MTLValueTransformerBlock? = { (value, success, error) in
        return dateFormatter.stringFromDate(value as! NSDate!)
    }
        
    return MTLValueTransformer(usingForwardBlock:forwardBlock, reverseBlock:reverseBlock)
}
```
*Original code from http://stackoverflow.com/a/29809466/5236819.*

- Changed both of my `Set<String>` properties to `Array<String>`.

*I don't know enough about how data structures are transformed between JSON and Swift to determine if using sets could be an issue or not.*

And then I came across this issue, #46, which is the same situation as mine. However, @matzew was using an older version of Mantle (late 2012) and the solution that worked for him has since been deprecated and removed from Mantle. 

[`mtl_externalRepresentationArrayTransformerWithModelClass:`](https://github.com/Mantle/Mantle/blob/master/Mantle/NSValueTransformer%2BMTLPredefinedTransformerAdditions.h#L30) has been removed. What replaced that function?

I believe this is my issue. 💀

**Comment by conradev** [comment]
Also, you can use `+[NSValueTransformer mtl_dateTransformerWithDateFormat:locale:]` for date transformations!

**Comment by conradev** [comment]
You will need `MTLJSONAdapter.arrayTransformerWithModelClass(:)` or `MTLJSONAdapter.dictionaryTransformerWithModelClass(:)` to transform sub-models (choosing one depending on whether or not there is a single model or an array of them).

**Comment by istx25** [comment]
@conradev: Xcode can't seem to find that function? Also, as mentioned in my iMessages, I found out that I must have a transformer for nested `MTLObject` objects but `JSONTransformerForKey(key:)` isn't called in `Observation`. Any ideas why?

**Comment by conradev** [comment]
Whatever contains the `Observation` objects does not have a transformer for them properly configured, it seems.

The error is occurring because the `Observation` model is not being transformed into anything, and it's not a valid JSON type.

**Comment by istx25** [comment]
@conradev and I moved over to iMessage so I could share the actual models with him instead of small snippets from here or there. We got it resolved, but I thought I would come back and share my learnings in case someone else comes around with a similar problem later!

The `"Invalid type in JSON write (GitHub.GHIssue)"` exception (usually) does not have to do with the `GHIssue` object itself, but which object contains `GHIssue` objects. In my case, I had a `Trial` object, which contained a Swift array of `Observation` objects (both subclasses of `MTLObject`).

```swift
// MARK: Property
var observations: [Observation]?

// MARK: Transform
static func observationsJSONTransformer() -> NSValueTransformer {
    return MTLJSONAdapter.arrayTransformerWithModelClass(Observation.self)
}
```
- Use `arrayTransformerWithModelClass` when in an array.

After resolving that issue, my problem still existed but in another area of my model layer, which Conrad helped me realise that I needed to do the same for all `MTLObject` submodels (even if they weren't in a collection).

```swift
// MARK: Property
var scene: TrialScene?

// MARK: Transform
static func sceneJSONTransformer() -> NSValueTransformer {
    return MTLJSONAdapter.dictionaryTransformerWithModelClass(TrialScene.self)
}
```
- Add one of these for each submodel/property (in every model that has submodels).
- Use `dictionaryTransformerWithModelClass` instead, since it's not in an array.

A lot of my confusion was caused because the documentation didn't explicitly explain/tell me how to set a transformer to a specific property/key in a model. Five transforms were defined (in the initial example) in the README, but only one (`updatedAtJSONTransformer`) was used in `JSONTransformerForKey` later on in the docs.

I was also confused by the following snippet:

>For added convenience, if you implement `+<key>JSONTransformer`, `MTLJSONAdapter` will use the result of that method instead.

At first glance (and Conrad explained this to me), I thought `<key>` had something to do with generics or something. I did not put the two together that the documentation meant replacing `<key>` with the name of the property/key (i.e.: `+somePropertyJSONTransformer`).

Food for thought, but I think these areas of the documentation could be improved. I'm quite busy right now, but when things slow back down, I'd love to contribute to the documentation to help make it a little easier to understand (and more explicit in these areas). Possibly even introducing hosted documentation for Mantle. Thoughts? I'll probably write up another issue about this soon.

But for now, thanks for all the help.

💃 😸
